File: app/src/main/java/de/herrmann_engel/rbv/activities/EditCardMedia.java
package de.herrmann_engel.rbv.activities;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.documentfile.provider.DocumentFile;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Objects;

import de.herrmann_engel.rbv.R;
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCard;
import de.herrmann_engel.rbv.databinding.ActivityEditCardMediaBinding;
import de.herrmann_engel.rbv.databinding.DiaFileExistsBinding;
import de.herrmann_engel.rbv.db.DB_Card;
import de.herrmann_engel.rbv.db.DB_Media;
import de.herrmann_engel.rbv.db.DB_Media_Link_Card;
import de.herrmann_engel.rbv.db.utils.DB_Helper_Create;
import de.herrmann_engel.rbv.db.utils.DB_Helper_Get;

public class EditCardMedia extends FileTools {
    private ActivityEditCardMediaBinding binding;
    private int cardNo;

    private DB_Helper_Get dbHelperGet;

    final ActivityResultLauncher<Intent> openFile = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    try {
                        Intent data = result.getData();
                        Uri fileUri = data.getData();
                        DocumentFile input = DocumentFile.fromSingleUri(this, fileUri);
                        DocumentFile outputDirectory = DocumentFile.fromTreeUri(this, Uri.parse(getCardMediaFolder()));
                        if (outputDirectory == null || !outputDirectory.isDirectory()) {
                            showSelectDialog(getResources().getString(R.string.select_folder_help_reselect));
                        } else if (input != null) {
                            String inputFileName = input.getName();
                            String mime = input.getType();
                            if (inputFileName != null && outputDirectory.findFile(inputFileName) == null) {
                                createMediaFile(outputDirectory, mime, inputFileName, fileUri);
                            } else {
                                Dialog fileExistsDialog = new Dialog(this, R.style.dia_view);
                                DiaFileExistsBinding bindingFileExistsDialog = DiaFileExistsBinding.inflate(getLayoutInflater());
                                fileExistsDialog.setContentView(bindingFileExistsDialog.getRoot());
                                fileExistsDialog.setTitle(getResources().getString(R.string.file_exists_title));
                                fileExistsDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                                        WindowManager.LayoutParams.MATCH_PARENT);

                                bindingFileExistsDialog.diaFileExistsLink.setOnClickListener(v -> {
                                    addMediaLink(inputFileName, mime);
                                    fileExistsDialog.dismiss();
                                });
                                bindingFileExistsDialog.diaFileExistsNew.setOnClickListener(v -> {
                                    createMediaFile(outputDirectory, mime, inputFileName, fileUri);
                                    fileExistsDialog.dismiss();
                                });
                                fileExistsDialog.show();
                            }
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                        Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show();
                    }
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityEditCardMediaBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        cardNo = getIntent().getExtras().getInt("card");
        dbHelperGet = new DB_Helper_Get(this);
        try {
            DB_Card card = dbHelperGet.getSingleCard(cardNo);
            TypedArray colors = getResources().obtainTypedArray(R.array.pack_color_main);
            TypedArray colorsBackground = getResources().obtainTypedArray(R.array.pack_color_background);
            int packColors = dbHelperGet.getSinglePack(card.pack).colors;
            if (packColors < Math.min(colors.length(), colorsBackground.length()) && packColors >= 0) {
                int color = colors.getColor(packColors, 0);
                int colorBackground = colorsBackground.getColor(packColors, 0);
                Objects.requireNonNull(getSupportActionBar()).setBackgroundDrawable(new ColorDrawable(color));
                Window window = this.getWindow();
                window.setStatusBarColor(color);
                binding.getRoot().setBackgroundColor(colorBackground);
            }
            colors.recycle();
            colorsBackground.recycle();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
        GradientDrawable gradientDrawable = new GradientDrawable();
        gradientDrawable.setShape(GradientDrawable.RECTANGLE);
        gradientDrawable.setColor(Color.argb(75, 200, 200, 250));
        gradientDrawable.setStroke(2, Color.rgb(170, 170, 220));
        gradientDrawable.setCornerRadius(8);
        binding.addMediaButton.setBackground(gradientDrawable);
        binding.addMediaButton.setOnClickListener(v -> {
            String folder = getCardMediaFolder();
            if (folder == null || folder.isEmpty()) {
                showSelectDialog(getResources().getString(R.string.select_folder_help));
            }
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("*/*");
            openFile.launch(intent);
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        setRecView();
    }

    private void createMediaFile(DocumentFile outputDirectory, String mime, String inputFileName, Uri fileUri) {
        try {
            DocumentFile output = outputDirectory.createFile(mime, inputFileName);
            String outputFileName = output.getName();
            OutputStream outputStream = getContentResolver().openOutputStream(output.getUri());
            InputStream inputStream = getContentResolver().openInputStream(fileUri);
            byte[] buffer = new byte[1024];
            int read;
            while ((read = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, read);
            }
            inputStream.close();
            outputStream.flush();
            outputStream.close();
            addMediaLink(outputFileName, mime);
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show();
        }
    }

    private void addMediaLink(String outputFileName, String mime) {
        try {
            DB_Helper_Create dbHelperCreate = new DB_Helper_Create(this);
            dbHelperCreate.createMedia(outputFileName, mime);
        } catch (Exception e) {
            e.printStackTrace();
        }
        try {
            if (dbHelperGet.existsMedia(outputFileName)) {
                DB_Media media = dbHelperGet.getSingleMedia(outputFileName);
                int fileId = media.uid;
                DB_Helper_Create dbHelperCreate = new DB_Helper_Create(this);
                dbHelperCreate.createMediaLink(fileId, cardNo);
                setRecView();
                Toast.makeText(this, R.string.success, Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show();
            }
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show();
        }
    }

    private void setRecView() {
        ArrayList<DB_Media_Link_Card> mediaList = (ArrayList<DB_Media_Link_Card>) dbHelperGet.getAllMediaLinksByCard(cardNo);
        AdapterMediaLinkCard adapter = new AdapterMediaLinkCard(mediaList, cardNo, getCardMediaFolder());
        binding.recCardMedia.setAdapter(adapter);
        binding.recCardMedia.setLayoutManager(new LinearLayoutManager(this));
    }

    @Override
    public void notifyFolderSet() {
        setRecView();
    }

    @Override
    protected void notifyMissingAction(int id) {
    }
}

File: app/src/main/java/de/herrmann_engel/rbv/activities/EditCardMedia.kt
package de.herrmann_engel.rbv.activities

import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.graphics.drawable.toDrawable
import androidx.documentfile.provider.DocumentFile
import androidx.recyclerview.widget.LinearLayoutManager
import de.herrmann_engel.rbv.R
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCard
import de.herrmann_engel.rbv.databinding.ActivityEditCardMediaBinding
import de.herrmann_engel.rbv.databinding.DiaFileExistsBinding
import de.herrmann_engel.rbv.db.DB_Media
import de.herrmann_engel.rbv.db.utils.DB_Helper_Create
import de.herrmann_engel.rbv.db.utils.DB_Helper_Get

class EditCardMedia : FileTools() {
    private lateinit var binding: ActivityEditCardMediaBinding
    private lateinit var dbHelperGet: DB_Helper_Get
    private var cardNo = 0
    private val openFile = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            try {
                val outputDirectory = DocumentFile.fromTreeUri(
                    this, Uri.parse(
                        cardMediaFolder
                    )
                )
                if (outputDirectory == null || !outputDirectory.isDirectory) {
                    showSelectDialog(resources.getString(R.string.select_folder_help_reselect))
                } else {
                    val data = result.data
                    val fileUri = data?.data
                    if (fileUri != null) {
                        val input = DocumentFile.fromSingleUri(this, fileUri)
                        if (input != null) {
                            val inputFileName = input.name
                            val mime = input.type
                            if (inputFileName != null && outputDirectory.findFile(inputFileName) == null) {
                                createMediaFile(outputDirectory, mime!!, inputFileName, fileUri)
                            } else {
                                val fileExistsDialog = Dialog(this, R.style.dia_view)
                                val bindingFileExistsDialog = DiaFileExistsBinding.inflate(
                                    layoutInflater
                                )
                                fileExistsDialog.setContentView(bindingFileExistsDialog.root)
                                fileExistsDialog.setTitle(resources.getString(R.string.file_exists_title))
                                fileExistsDialog.window!!.setLayout(
                                    WindowManager.LayoutParams.MATCH_PARENT,
                                    WindowManager.LayoutParams.MATCH_PARENT
                                )
                                bindingFileExistsDialog.diaFileExistsLink.setOnClickListener {
                                    addMediaLink(inputFileName, mime)
                                    fileExistsDialog.dismiss()
                                }
                                bindingFileExistsDialog.diaFileExistsNew.setOnClickListener {
                                    createMediaFile(
                                        outputDirectory,
                                        mime!!,
                                        inputFileName!!,
                                        fileUri
                                    )
                                    fileExistsDialog.dismiss()
                                }
                                fileExistsDialog.show()
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEditCardMediaBinding.inflate(layoutInflater)
        setContentView(binding.root)
        dbHelperGet = DB_Helper_Get(this)
        cardNo = intent.extras!!.getInt("card")
        val card = dbHelperGet.getSingleCard(cardNo)
        val colorsStatusBar = resources.obtainTypedArray(R.array.pack_color_statusbar)
        val colorsBackground = resources.obtainTypedArray(R.array.pack_color_background)
        val minimalLength = colorsStatusBar.length().coerceAtMost(colorsBackground.length())
        val packColors = dbHelperGet.getSinglePack(card.pack).colors
        if (packColors in 0..<minimalLength) {
            val colorStatusBar = colorsStatusBar.getColor(packColors, 0)
            val colorBackground = colorsBackground.getColor(packColors, 0)
            supportActionBar?.setBackgroundDrawable(colorStatusBar.toDrawable())
            binding.root.setBackgroundColor(colorBackground)
        }
        colorsStatusBar.recycle()
        colorsBackground.recycle()
        binding.addMediaButton.setOnClickListener {
            val folder = cardMediaFolder
            if (folder.isNullOrEmpty()) {
                showSelectDialog(resources.getString(R.string.select_folder_help))
            } else {
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "*/*"
                openFile.launch(intent)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        setRecView()
    }

    private fun createMediaFile(
        outputDirectory: DocumentFile,
        mime: String,
        inputFileName: String,
        fileUri: Uri
    ) {
        try {
            val output = outputDirectory.createFile(mime, inputFileName)!!
            val outputFileName = output.name
            val outputStream = contentResolver.openOutputStream(output.uri)!!
            val inputStream = contentResolver.openInputStream(fileUri)!!
            val buffer = ByteArray(1024)
            var read: Int
            while (inputStream.read(buffer).also { read = it } != -1) {
                outputStream.write(buffer, 0, read)
            }
            inputStream.close()
            outputStream.flush()
            outputStream.close()
            addMediaLink(outputFileName, mime)
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show()
        }
    }

    private fun addMediaLink(outputFileName: String?, mime: String?) {
        try {
            val dbHelperCreate = DB_Helper_Create(this)
            if (!dbHelperGet.existsMedia(outputFileName)) {
                dbHelperCreate.createMedia(outputFileName, mime)
            }
            if (dbHelperGet.existsMedia(outputFileName)) {
                val media = dbHelperGet.getSingleMedia(outputFileName)
                val fileId = media.uid
                dbHelperCreate.createMediaLink(fileId, cardNo)
                setRecView()
                Toast.makeText(this, R.string.success, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show()
        }
    }

    private fun setRecView() {
        val mediaList = dbHelperGet.getCardMedia(cardNo) as ArrayList<DB_Media>
        val adapter = AdapterMediaLinkCard(mediaList, cardNo, cardMediaFolder)
        binding.recCardMedia.adapter = adapter
        binding.recCardMedia.layoutManager = LinearLayoutManager(this)
    }

    public override fun notifyFolderSet() {
        setRecView()
    }

    override fun notifyMissingAction(id: Int) {}
}
